### ISSUE TITLE
Question: version integration with Git tags?

### ISSUE THREAD

**Issue by jamesbraza** [citation-file-format/citation-file-format#435]
Is it possible to integrate the `version` field of a `CITATION.cff` with Git tags (or GitHub releases)?

I don't see any mentions of tools to help with this in https://github.com/citation-file-format/citation-file-format#tools-to-work-with-citationcff-files-wrench, the closest is probably the GitHub bot from https://github.com/citation-file-format/citation-file-format/issues/238.

Disclaimer: coming from [here](https://github.com/nipy/nibabel/issues/1230#issuecomment-1546792463).

**Comment by jspaaks** [comment]
Hi James, thanks for making the issue. Just to clarify, you're asking whether it's possible to automatically update the value of the `version` key in `CITATION.cff` files when tagging a commit or making a release?

I'm not aware of any tool that does this. Some related tools in this space that may form part of a solution:

1. versioneer https://pypi.org/project/versioneer/
1. bump2version https://pypi.org/project/bump2version/
1. yq https://pypi.org/project/yq/
1. precommit https://pypi.org/project/pre-commit/, https://pre-commit.com/

**Comment by kevinmatthes** [comment]
@jamesbraza

Yes, it is possible to integrate the version update -- not only of a CITATION.cff -- with GitHub releases.  Therefore, first of all, you need to choose an assisting software which will increment your versions.  @jspaaks already named some of them; in the context of Rust, there is also `cargo-release` available.  Which tool fits your needs best depends on your exact use case.

The process of configuration of such a tool involves, as the next step, the identification of all files which require a version increment on release.  To do so, use `git grep` and the current version of your project to locate all files.  For example, you would search for version 1.2.3 of your software by `git grep 1\.2\.3`.  When you need to update the version of a CITATION.cff, you might need to update other files containing the version as a magic number, as well.

Then, you add all found files to the replacement rules of the version incrementing system you chose.  I previously worked with primarily with bump2version but `cargo-release` has a similar approach:

- the replacement rules are defined per file.
- you need / are allowed to create searching hooks; `cargo-release` requires -- as far as I know -- exact instructions to work, bump2version also works with just a file name given and will then replace all occurrences of the configured version number.

Finally, you integrate the system you chose into the release workflow.  This can happen either literally by adding calls to the tool to the GitHub Action workflow definition or you just remember to have to call the tool when preparing a release by hand.

Besides the `version` field which can be updated rather easily by a version incrementing system, you might also need to take care about the `date-released` entry.  You can do so by running a `sed` call or use the Action I wrote to do so:  https://github.com/kevinmatthes/cff-release-today.  (You cannot find it in the table because the request to add it to the table is still pending; #427.)  If you should decide to use bump2version, there is also an Action which incorporates the `date-released` update Action to call bump2version, as well:  https://github.com/kevinmatthes/release-bump2version-scriv.  As the name already indicates, you need to set up Scriv, as well, which will then assemble your CHANGELOG on release, based on the fragments your contributors added by their Pull Requests.

If you need help with the setup of bump2version, Scriv, and / or one the Actions, I can help you, if required.

**Comment by jamesbraza** [comment]
> Just to clarify, you're asking whether it's possible to automatically update the value of the `version` key in `CITATION.cff` files when tagging a commit or making a release

Yes @jspaaks you interpreted my question correctly!

And thank you @kevinmatthes for the detailed answer too, I gave [cff-release-today](https://github.com/kevinmatthes/cff-release-today) a GitHub star as thanks 😄 

--- 

My takeaways here are:
- `CITATION.cff` itself is just a text file that doesn't have "integrations"
- One can use automated tooling (and thank you both for suggesting great options) to sync the `version` with a Git tag

I believe my question has been answered, so please feel free to close this out.

---

**Alternate Solution**

It might be cool if `CITATION.cff` somehow featured a magic value (e.g. `@git-tag`) for the `version` value that GitHub could then know to auto-populate the latest Git tag, when one clicks the "Cite this repository" dropdown.  Just sharing an idea of how one could rely on the repository host to do this

![image](https://github.com/citation-file-format/citation-file-format/assets/8990777/527b6a24-693f-41db-81de-fe6a337a82a4)

**Comment by apirogov** [comment]
We are building a tool called [somesy](https://github.com/Materials-Data-Science-and-Informatics/somesy), it does not integrate with git directly, but with software-specific metadata formats such as `pyproject.toml` and `package.json`. Maybe that could be of interest. It also synchronizes much more than version. If what you want is basically mirroring your project metadata across various relevant formats flying around in a repo (including CITATION.cff), maybe take a look.

**Comment by sdruskat** [comment]
> It might be cool if `CITATION.cff` somehow featured a magic value (e.g. `@git-tag`) for the `version` value that GitHub could then know to auto-populate the latest Git tag, when one clicks the "Cite this repository" dropdown. Just sharing an idea of how one could rely on the repository host to do this
> 
> ![image](https://user-images.githubusercontent.com/8990777/238424484-527b6a24-693f-41db-81de-fe6a337a82a4.png)

We had a similar idea in the past (#197), but have originally decided to not implement this. Feel free to discuss this approach specifically in the re-opened issue.
